Can't get lesbians. Help!

```
$ reddit-dl -url https://www.reddit.com/r/Lesbian_gifs/comments/zvx95q/our_favourite_position_is_69
[INFO]: Parsing The URL
[INFO]: Getting the JSON body
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1 [running]:
github.com/noornee/reddit-dl/utility.ParseJSONBody({0xc000258500, 0x23bd, 0x2500})
        /git/reddit-dl/utility/utility.go:111 +0x7f7
main.main()
        /git/reddit-dl/main.go:42 +0x219
```

--- a/cmd/reddit-dl/main.go
+++ b/cmd/reddit-dl/main.go
@@ -19,6 +19,13 @@ func main() {
 	passed := utility.IsFlagPassed(raw_url)
 
 	// if a flag isnt passed and a url is passed, assign the url to raw_url
+
+	/*
+		perhaps this?
+		if !(len(raw_url) > 0) {
+			...
+		}
+	*/
 	if len(os.Args) > 1 && !strings.HasSuffix(os.Args[1], "-url") {
 		raw_url = os.Args[1]
 	} else {
@@ -41,7 +48,7 @@ func main() {
 	media_url, err := utility.ParseJSONBody(body)
 	if err != nil {
 		utility.ErrorLog.Fatal(err)
-		return
+		// return Fatal calls Print, then os.Exit(1), can't return.
 	}
 	//fmt.Println(media_url)
 

--- a/handler/handler.go
+++ b/handler/handler.go
@@ -4,27 +4,22 @@ import (
 	"fmt"
 	"io"
 	"net/http"
-	"regexp"
 	"strings"
 
 	"github.com/noornee/reddit-dl/utility"
 )
-
-func check(err error) {
-	if err != nil {
-		utility.ErrorLog.Println(err)
-	}
-}
 
 // Parses the url and appends .json to it
 func ParseUrl(raw string) (url, title string) {
 
 	utility.InfoLog.Println("Parsing The URL")
 
-	re, _ := regexp.Compile("www")
+	//re, _ := regexp.Compile("www")
 
 	// i.e. 'https://www.reddit.com/...' -> 'https://old.reddit.com/...'
-	subdomain := re.ReplaceAllString(raw, "old")
+	//subdomain := re.ReplaceAllString(raw, "old")
+	// strings.Replace is lighter than regexp
+	subdomain := strings.Replace(raw, "www.reddit.com", "old.reddit.com", -1)
 
 	split_url := strings.Split(subdomain, "?")[0]
 
@@ -45,7 +40,9 @@ func ParseUrl(raw string) (url, title string) {
 func GetBody(url string) ([]byte, error) {
 
 	req, err := http.NewRequest("GET", url, nil)
-	check(err)
+	if err != nil {
+		return nil, err
+	}
 
 	req.Header.Set("User-Agent", "Mozilla/5.0")
 
@@ -60,7 +57,9 @@ func GetBody(url string) ([]byte, error) {
 	defer resp.Body.Close()
 
 	body, err := io.ReadAll(resp.Body)
-	check(err)
+	if err != nil {
+		return nil, err
+	}
 
 	return body, nil
 
@@ -69,10 +68,14 @@ func GetBody(url string) ([]byte, error) {
 func GetHead(url string) (status_code int, content_type string) {
 
 	req, err := http.NewRequest("HEAD", url, nil)
-	check(err)
+	if err != nil {
+		return 0, ""
+	}
 
 	resp, err := http.DefaultClient.Do(req)
-	check(err)
+	if err != nil {
+		return 0, ""
+	}
 
 	defer resp.Body.Close()
 

--- a/service/downloader.go
+++ b/service/downloader.go
@@ -103,7 +103,7 @@ Loop:
 		file_name := fmt.Sprintf("%s.mp4", title)
 		err := os.Rename(resp.Filename, file_name)
 		if err != nil {
-			fmt.Println(err)
+			utility.ErrorLog.Println(err)
 		}
 
 		utility.InfoLog.Printf("Download saved to %v \n", file_name)

--- a/service/ffmpeg.go
+++ b/service/ffmpeg.go
@@ -1,8 +1,6 @@
 package service
 
 import (
-	"fmt"
-	"io/ioutil"
 	"os"
 
 	"github.com/noornee/reddit-dl/utility"
@@ -16,7 +14,8 @@ func video_merger(filename string) {
 
 	filename = filename + ".mp4"
 
-	files, err := ioutil.ReadDir(temp_dir)
+	// use os.ReadDir instead of ioutil, as it's deprecated since go 1.16
+	files, err := os.ReadDir(temp_dir)
 
 	if err != nil {
 		utility.ErrorLog.Println(err)
@@ -33,8 +32,7 @@ func video_merger(filename string) {
 	in1 := ffmpeg.Input(vid)
 	in2 := ffmpeg.Input(aud)
 
-	utility.InfoLog.Printf("Merging files into \t%s", filename)
-	fmt.Println()
+	utility.InfoLog.Printf("Merging files into \t%s\r\n", filename)
 
 	err = ffmpeg.Concat([]*ffmpeg.Stream{in1, in2}, ffmpeg.KwArgs{"v": 1, "a": 1}).
 		Output(filename, ffmpeg.KwArgs{"v": "quiet"}).

--- a/utility/utility.go
+++ b/utility/utility.go
@@ -38,10 +38,11 @@ func ParseJSONBody(file []byte) ([]string, error) {
 
 	// -------------------------------------BASE------------------------------------------------------------ //
 	root, ok := dataDump.([]any)
-	if ok != true {
+	if !ok {
 		return urls, errors.New("cannot parse body")
 	}
 
+	// use a struct.. this is really, really hard to read & use!
 	edge := root[0].(hashMap)
 	data := edge["data"].(hashMap)
 	children := data["children"].([]any)
@@ -56,8 +57,8 @@ func ParseJSONBody(file []byte) ([]string, error) {
 			media_id := metadata[i].(hashMap)
 			media_s := media_id["s"].(hashMap)
 			media_url := media_s["u"]
-			new_media_url := strings.ReplaceAll(fmt.Sprint(media_url), "amp;", "")
-			urls = append(urls, fmt.Sprint(new_media_url))
+			new_media_url := strings.ReplaceAll(media_url.(string), "amp;", "")
+			urls = append(urls, new_media_url)
 		}
 		return urls, nil
 	}
@@ -68,22 +69,35 @@ func ParseJSONBody(file []byte) ([]string, error) {
 	if secure_media == nil {
 		// for normal image/gif
 		url_overridden_by_dest := data2["url_overridden_by_dest"]
-		urls = append(urls, fmt.Sprint(url_overridden_by_dest))
+		urls = append(urls, url_overridden_by_dest.(string))
 		return urls, nil
 	}
 
 	// ----------------------------------------CROSSPOST--------------------------------------------------- //
 	// if it doesn't have the underlying interface `ok` would be false then its a crosspost
 	// for reddit cross post video
-	if ok != true {
+	if !ok {
 		cross_post := data2["crosspost_parent_list"].([]any)
 		data3 := cross_post[0].(hashMap)
 		secure_media := data3["secure_media"].(hashMap)
 		reddit_video := secure_media["reddit_video"].(hashMap)
 		fallback_url := reddit_video["fallback_url"]
-		urls = append(urls, fmt.Sprint(fallback_url))
+		urls = append(urls, fallback_url.(string))
 		return urls, nil
 
+	}
+
+	// --------------------------------"PREVIEW" VIDEOS--------------------------------------------------- //
+	preview := data2["preview"].(hashMap)
+
+	reddit_video_preview := preview["reddit_video_preview"].(hashMap)
+
+	fallbackurl, ok := reddit_video_preview["fallback_url"]
+
+	if ok {
+		urls = append(urls, fallbackurl.(string))
+
+		return urls, nil
 	}
 
 	// --------------------------------FOR GIFS/VIDEO HOSTED ON GFYCAT.COM----------------------------------- //
@@ -92,19 +106,19 @@ func ParseJSONBody(file []byte) ([]string, error) {
 		provider_url := oembed["provider_url"]
 		thumbnail_url := oembed["thumbnail_url"]
 		if provider_url == "https://gfycat.com" {
-			new_url := strings.ReplaceAll(fmt.Sprint(thumbnail_url), "size_restricted.gif", "mobile.mp4")
-			urls = append(urls, fmt.Sprint(new_url))
+			new_url := strings.ReplaceAll(thumbnail_url.(string), "size_restricted.gif", "mobile.mp4")
+			urls = append(urls, new_url)
 			return urls, nil
 
 		} else {
-			return urls, errors.New(fmt.Sprintf("unsupported provider \"%s\"", provider_url))
+			return urls, fmt.Errorf("unsupported provider \"%s\"", provider_url)
 		}
 	}
 
 	// --------------------------------NORMAL REDDIT VIDEO------------------------------------------------- //
 	reddit_video := secure_media["reddit_video"].(hashMap)
 	fallback_url := reddit_video["fallback_url"]
-	urls = append(urls, fmt.Sprint(fallback_url))
+	urls = append(urls, fallback_url.(string))
 	return urls, nil
 
 	// ---------------------------------------------------------------------------------------------------- //

